File: wegent-evaluate/frontend/src/apis/analytics.ts
/**
 * API client for analytics endpoints
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL || 'http://localhost:18000'

export async function getTrends(params: {
  start_date: string
  end_date: string
  metric?: 'faithfulness' | 'answer_relevancy' | 'context_precision' | 'overall'
  group_by?: 'day' | 'week' | 'month'
  retriever_name?: string
  embedding_model?: string
  version_id?: number
}) {
  const searchParams = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null) {
      searchParams.append(key, String(value))
    }
  })

  const response = await fetch(
    `${API_BASE_URL}/api/analytics/trends?${searchParams.toString()}`
  )
  if (!response.ok) throw new Error('Failed to get trends')
  return response.json()
}

export async function getRetrieverComparison(params: {
  start_date: string
  end_date: string
  version_id?: number
}) {
  const searchParams = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null) {
      searchParams.append(key, String(value))
    }
  })

  const response = await fetch(
    `${API_BASE_URL}/api/analytics/comparison/retriever?${searchParams.toString()}`
  )
  if (!response.ok) throw new Error('Failed to get retriever comparison')
  return response.json()
}

export async function getEmbeddingComparison(params: {
  start_date: string
  end_date: string
  version_id?: number
}) {
  const searchParams = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null) {
      searchParams.append(key, String(value))
    }
  })

  const response = await fetch(
    `${API_BASE_URL}/api/analytics/comparison/embedding?${searchParams.toString()}`
  )
  if (!response.ok) throw new Error('Failed to get embedding comparison')
  return response.json()
}

export async function getContextComparison(subtaskContextId: number) {
  const response = await fetch(
    `${API_BASE_URL}/api/analytics/comparison/context/${subtaskContextId}`
  )
  if (!response.ok) throw new Error('Failed to get context comparison')
  return response.json()
}

export async function getIssuesAnalytics(params: {
  start_date: string
  end_date: string
  version_id?: number
}) {
  const searchParams = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null) {
      searchParams.append(key, String(value))
    }
  })

  const response = await fetch(
    `${API_BASE_URL}/api/analytics/issues?${searchParams.toString()}`
  )
  if (!response.ok) throw new Error('Failed to get issues analytics')
  return response.json()
}

File: wegent-evaluate/backend/app/api/endpoints/analytics.py
"""
Analytics API endpoints.
"""
from datetime import datetime
from typing import Literal, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db
from app.schemas.analytics import (
    ContextComparisonResponse,
    EmbeddingComparisonResponse,
    IssuesAnalyticsResponse,
    RetrieverComparisonResponse,
    TrendsResponse,
)
from app.services.analytics_service import AnalyticsService

router = APIRouter()


@router.get("/trends", response_model=TrendsResponse)
async def get_trends(
    start_date: datetime,
    end_date: datetime,
    metric: Literal[
        "faithfulness", "answer_relevancy", "context_precision", "overall"
    ] = "overall",
    group_by: Literal["day", "week", "month"] = "day",
    retriever_name: Optional[str] = None,
    embedding_model: Optional[str] = None,
    version_id: Optional[int] = Query(None, description="Filter by version ID"),
    db: AsyncSession = Depends(get_db),
):
    """Get trend data for a specific metric over time."""
    service = AnalyticsService(db)
    data = await service.get_trends(
        start_date=start_date,
        end_date=end_date,
        metric=metric,
        group_by=group_by,
        retriever_name=retriever_name,
        embedding_model=embedding_model,
        version_id=version_id,
    )

    return TrendsResponse(
        metric=metric,
        group_by=group_by,
        data=data,
    )


@router.get("/comparison/retriever", response_model=RetrieverComparisonResponse)
async def get_retriever_comparison(
    start_date: datetime,
    end_date: datetime,
    version_id: Optional[int] = Query(None, description="Filter by version ID"),
    db: AsyncSession = Depends(get_db),
):
    """Compare evaluation metrics across different retrievers."""
    service = AnalyticsService(db)
    data = await service.get_retriever_comparison(
        start_date=start_date,
        end_date=end_date,
        version_id=version_id,
    )

    return RetrieverComparisonResponse(data=data)


@router.get("/comparison/embedding", response_model=EmbeddingComparisonResponse)
async def get_embedding_comparison(
    start_date: datetime,
    end_date: datetime,
    version_id: Optional[int] = Query(None, description="Filter by version ID"),
    db: AsyncSession = Depends(get_db),
):
    """Compare evaluation metrics across different embedding models."""
    service = AnalyticsService(db)
    data = await service.get_embedding_comparison(
        start_date=start_date,
        end_date=end_date,
        version_id=version_id,
    )

    return EmbeddingComparisonResponse(data=data)


@router.get("/comparison/context/{subtask_context_id}", response_model=ContextComparisonResponse)
async def get_context_comparison(
    subtask_context_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Get comparison of all evaluations for a specific context."""
    service = AnalyticsService(db)
    data = await service.get_context_comparison(subtask_context_id=subtask_context_id)

    return ContextComparisonResponse(**data)


@router.get("/issues", response_model=IssuesAnalyticsResponse)
async def get_issues_analytics(
    start_date: datetime,
    end_date: datetime,
    version_id: Optional[int] = Query(None, description="Filter by version ID"),
    db: AsyncSession = Depends(get_db),
):
    """Get analytics on issue types."""
    service = AnalyticsService(db)
    data = await service.get_issues_analytics(
        start_date=start_date,
        end_date=end_date,
        version_id=version_id,
    )

    return IssuesAnalyticsResponse(**data)
